File: src/decoders/assignment.ts
import { type Assignment, type SessionHandle, AssignmentReturnKind } from "~/models";
import { decodeSubject } from "./subject";
import { decodePronoteDate } from "./pronote-date";
import { decodeAttachment } from "./attachment";
import { decodeAssignmentTheme } from "./assignment-theme";

export const decodeAssignment = (assignment: any, session: SessionHandle): Assignment => {
  return {
    id: assignment.N,
    subject: decodeSubject(assignment.Matiere.V),
    description: assignment.descriptif.V,
    backgroundColor: assignment.CouleurFond,
    done: assignment.TAFFait,
    deadline: decodePronoteDate(assignment.PourLe.V),
    attachments: assignment.ListePieceJointe.V.map((attachment: any) => decodeAttachment(attachment, session)),
    difficulty: assignment.niveauDifficulte,
    length: assignment.duree,
    themes: assignment.ListeThemes.V.map((theme: any) => decodeAssignmentTheme(theme)),
    return: {
      kind: assignment.genreRendu ?? AssignmentReturnKind.None,
      canUpload: assignment.peuRendre ?? false,
      uploaded: assignment.documentRendu && decodeAttachment(assignment.documentRendu.V, session)
    },
    lessonResourceID: assignment.cahierDeTextes?.V.N
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/assignment.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.*

import kotlinx.serialization.json.*

fun decodeAssignment (assignment: JsonObject, sessionInformation: SessionInformation) = Assignment(
    id = assignment["N"]!!.jsonPrimitive.content,
    subject = decodeSubject(assignment["Matiere"]!!.jsonObject["V"]!!.jsonObject),
    description = assignment["descriptif"]!!.jsonObject["V"]!!.jsonPrimitive.content,
    backgroundColor = assignment["CouleurFond"]!!.jsonPrimitive.content,
    done = assignment["TAFFait"]!!.jsonPrimitive.boolean,
    deadline = decodePronoteDate(assignment["PourLe"]!!.jsonObject["V"]!!.jsonPrimitive.content),
    attachments = assignment["ListePieceJointe"]!!.jsonObject["V"]!!.jsonArray.map { decodeAttachment(it.jsonObject, sessionInformation) },
    difficulty = AssignmentDifficulty.fromInt(assignment["niveauDifficulte"]!!.jsonPrimitive.int),
    length = assignment["duree"]!!.jsonPrimitive.double,
    themes = assignment["ListeThemes"]!!.jsonObject["V"]!!.jsonArray.map { decodeAssignmentTheme(it.jsonObject) },
    `return` = AssignmentReturn(
        kind = if (assignment["genreRendu"] != null) AssignmentReturnKind.fromInt(assignment["genreRendu"]!!.jsonPrimitive.int) else AssignmentReturnKind.None,
        canUpload = assignment["peuRendre"]?.jsonPrimitive?.boolean ?: false,
        uploaded = if (assignment["documentRendu"] != null) decodeAttachment(assignment["documentRendu"]!!.jsonObject["V"]!!.jsonObject, sessionInformation) else null
    ),
    resourceID = if (assignment["cahierDeTextes"] != null) assignment["cahierDeTextes"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content else null
)